pxr/base/tf/stringUtils.h

```diff
@@ -683,18 +727,113 @@ std::string TfStringCatPaths( const std::string &prefix,
 inline bool
 TfIsValidIdentifier(std::string const &identifier)
 {
-    char const *p = identifier.c_str();
-    auto letter = [](unsigned c) { return ((c-'A') < 26) || ((c-'a') < 26); };
-    auto number = [](unsigned c) { return (c-'0') < 10; };
-    auto under = [](unsigned c) { return c == '_'; };
-    unsigned x = *p;
-    if (!x || number(x)) {
-        return false;
+    if (UseUTF8Identifiers())
+    {
+        // use unicode utils to validate the identifier
+        return TfUnicodeUtils::IsValidUTF8Identifier(identifier, 
+            identifier.begin(), identifier.end());
+    }
+    else
+    {
+        char const *p = identifier.c_str();
+        auto letter = [](unsigned c) { return ((c-'A') < 26) || ((c-'a') < 26); };
+        auto number = [](unsigned c) { return (c-'0') < 10; };
+        auto under = [](unsigned c) { return c == '_'; };
+        unsigned x = *p;
+        if (!x || number(x)) {
+            return false;
+        }
+        while (letter(x) || number(x) || under(x)) {
+            x = *p++;
+        };
+        return x == 0;
     }
-    while (letter(x) || number(x) || under(x)) {
-        x = *p++;
-    };
-    return x == 0;
+}
+
+/// Test whether the subsequence of \a identifier given by begin and end 
+/// is a valid identifier name.
+///
+/// An identifier name is valid if it consists of any valid UTF-8 sequence 
+/// that is at least one character long and fulfills the standard C/Python 
+/// identifier convention for Unicode (XID_Start followed by 0 or more 
+/// XID_Continue class characters).  If the TF_UTF8_IDENTIFIERS environment 
+/// setting is overridden to false, identifier validity is checked against 
+/// legacy rules and follows the C/Python identifier convention for ASCII 
+/// characters (i.e. at least one character long, must start with a letter 
+/// or underscore, and must contain only letters, underscores, and numerals).
+/// WARNING: This method is for internal use only and performs no 
+///          validation on begin / end.
+/// 
+inline bool
+_TfIsValidIdentifierSubsequence(const std::string& identifier,
+    std::string::const_iterator begin, std::string::const_iterator end)
+{
+    if (UseUTF8Identifiers())
+    {
+        // use unicode utils to validate the identifier
+        return TfUnicodeUtils::IsValidUTF8Identifier(identifier, 
+            begin, end);
+    }
+    else
+    {
+        size_t count = (end - begin);
+        char const *p = &(*begin);
+        auto letter = [](unsigned c) { return ((c-'A') < 26) || ((c-'a') < 26); };
```

Comment:
Since some of these checks are shared with TfIsValidIdentifier above, it might be nice to break them out into a shared inlined function (like `IsALetter` or something)? I get paranoid about drift in the future in case someone edited one and forgot the other etc...

Reply:
The original change did exactly that, and the white paper describes that issue.  For this PR, I intentionally left all code paths that were there the same, so as not to introduce anything extra.